lang/syn/src/codegen/accounts/constraints.rs

```diff
@@ -320,6 +325,56 @@ pub fn generate_constraint_rent_exempt(
     }
 }
 
+fn generate_constraint_realloc(f: &Field, c: &ConstraintReallocGroup) -> proc_macro2::TokenStream {
+    let field = &f.ident;
+    let account_name = field.to_string();
+    let new_space = &c.space;
+    let payer = &c.payer;
+    let zero = &c.zero;
+
+    quote! {
+        if __reallocs.contains(&#field.key().to_string()) {
+            return Err(anchor_lang::error::Error::from(anchor_lang::error::ErrorCode::AccountDuplicateReallocs).with_account_name(#account_name));
+        }
+
+        let __anchor_rent = anchor_lang::prelude::Rent::get()?;
+        let __field_info = #field.to_account_info();
+        let __additive = #new_space > __field_info.data_len();
+
+        let __delta_space = if __additive {
+            #new_space.checked_sub(__field_info.data_len()).unwrap()
+        } else {
+            __field_info.data_len().checked_sub(#new_space).unwrap()
+        };
+
+        if __delta_space > 0 {
+            if __delta_space > anchor_lang::solana_program::entrypoint::MAX_PERMITTED_DATA_INCREASE {
```

Comment:
Should this constraint only apply if `__additive`?

Reply:
i've gone back and forth on this. most use-cases are going to be for scaling out larger accounts, but can see why it would be useful to do account size/data reduction as well.

@armaniferrante thoughts?

Reply:
I believe the issue is that since `__delta_space` is always positive, this would incorrectly prevent shrinking an account by more than `MAX_PERMITTED_DATA_INCREASE `

Reply:
@ethanwu10 are you saying the check should be:
```rs
if __additive && __delta_bytes > anchor_lang::solana_program::... {}
```

Reply:
it is a max _increase_ limit 🤔 and docs don't specify anything about reductive reallocation limitations.

Reply:
Yes, I think the check above would make more sense.

Reply:
will change in next commit.

Reply:
I guess another question while we're at it, why not keep `__delta_space` as a signed number instead of dealing with an `__additive` flag? There's no way anyone is making accounts large enough to overflow an isize, so `.try_into().unwrap()` seems perfectly reasonable to me...

Reply:
so switching the calculation to this?
```rs
let __delta_bytes: isize = (#new_space as isize).checked_sub(__field_info.data_len() as isize).unwrap();
```

Reply:
This would silently wrap sizes that are out of range for an `isize`. My intuition says it's *probably* fine, but that at least warrants a comment justifying the `as` cast. The safer way is
```rust
(::std::convert::TryInto::<isize>::try_into(#new_space).unwrap()).checked_sub(::std::convert::TryInto::try_into(__field_info.data_len()).unwrap()).unwrap()

Reply:
i had originally kept them all as `usize` with the `__additive` flag for easy compatibility with existing APIs, and comparing to `MAX_PERMITTED_DATA_INCREASE` without lots of conversions.